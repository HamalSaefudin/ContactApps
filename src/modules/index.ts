export {default as HomeScreen} from './Home/HomeScreen';
export {default as DetailContactSCreen} from './DetailContact/DetailContactScreen';
