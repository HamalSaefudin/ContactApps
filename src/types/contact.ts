export interface IContact {
  firstName: string;
  lastName: string;
  age: number;
  photo: string;
  id: string;
  callHistory?: string;
}

export interface ICreateContact {
  firstName: string;
  lastName: string;
  age: number;
  photo: string;
}
