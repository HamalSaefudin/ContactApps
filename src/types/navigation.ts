export enum routesEnum {
  HOME = 'HOME',
  DETAIL_CONTACT = 'DETAIL_CONTACT',
}

export type RoutesParamList = {
  [routesEnum.HOME]: undefined;
  [routesEnum.DETAIL_CONTACT]: {id: string};
};
