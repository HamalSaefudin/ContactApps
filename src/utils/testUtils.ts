import {Reducer, createStore} from 'redux';

export const globalMockReducers = (store: Reducer) => {
  return createStore(store);
};
