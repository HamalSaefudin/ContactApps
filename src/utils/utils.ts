export const getImg = (url: string): string => {
  if (url.startsWith('http://') || url.startsWith('https://')) {
    return url.startsWith('http://') ? url.replace('http://', 'https://') : url;
  } else {
    return 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQ4YreOWfDX3kK-QLAbAL4ufCPc84ol2MA8Xg&s';
  }
};
