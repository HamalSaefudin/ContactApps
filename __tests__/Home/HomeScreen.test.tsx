import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import {QueryClient, QueryClientProvider} from 'react-query';
import {globalMockReducers} from '@/utils/testUtils';
import {combineReducers} from 'redux';
import {HomeScreen} from '@/modules';
import {routesEnum} from '@/types/navigation';

const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: mockNavigate,
  }),
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('react-query', () => ({
  ...jest.requireActual('react-query'),
  useQuery: jest.fn().mockImplementation(() => ({
    data: [
      {
        id: '1',
        firstName: 'John',
        lastName: 'Doe',
        photo: 'https://randomuser.me/api/portraits/men/1.jpg',
        age: 30,
        callHistory: '2024-06-09 12:00:00',
      },
    ],
    isFetching: false,
  })),
}));

describe('HomeScreen', () => {
  let store: any;

  beforeEach(() => {
    store = globalMockReducers(
      combineReducers({
        contacts: () => ({
          lastVisitContact: [
            {
              id: '2',
              firstName: 'Jane',
              lastName: 'Doe',
              photo: 'https://randomuser.me/api/portraits/women/1.jpg',
              age: 25,
              callHistory: '2024-06-09 14:00:00',
            },
          ],
        }),
      }),
    );
  });

  it('to match snapshot', () => {
    const queryClient = new QueryClient();

    const {toJSON} = render(
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <NavigationContainer>
            <HomeScreen
              navigation={{navigate: mockNavigate} as any}
              route={{
                params: undefined,
                path: '',
                key: '',
                name: routesEnum.HOME,
              }}
            />
          </NavigationContainer>
        </QueryClientProvider>
      </Provider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly and navigate', () => {
    const queryClient = new QueryClient();

    const {getByText} = render(
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <NavigationContainer>
            <HomeScreen
              navigation={{navigate: mockNavigate} as any}
              route={{
                params: undefined,
                path: '',
                key: '',
                name: routesEnum.HOME,
              }}
            />
          </NavigationContainer>
        </QueryClientProvider>
      </Provider>,
    );

    expect(getByText('John Doe')).toBeTruthy();

    fireEvent.press(getByText('John Doe'));

    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'SET_LAST_VISIT_CONTACT',
      payload: {
        id: '1',
        firstName: 'John',
        lastName: 'Doe',
        photo: 'https://randomuser.me/api/portraits/men/1.jpg',
        age: 30,
        callHistory: '2024-06-09 12:00:00',
      },
    });
    expect(mockNavigate).toHaveBeenCalledWith(routesEnum.DETAIL_CONTACT, {
      id: '1',
    });
  });

  it('renders correctly and navigate to screen detail contact', () => {
    const queryClient = new QueryClient();

    const {getByTestId} = render(
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <NavigationContainer>
            <HomeScreen
              navigation={{navigate: mockNavigate} as any}
              route={{
                params: undefined,
                path: '',
                key: '',
                name: routesEnum.HOME,
              }}
            />
          </NavigationContainer>
        </QueryClientProvider>
      </Provider>,
    );

    fireEvent.press(getByTestId('add-button'));

    expect(mockNavigate).toHaveBeenCalledWith(routesEnum.DETAIL_CONTACT, {
      id: '',
    });
  });
});
