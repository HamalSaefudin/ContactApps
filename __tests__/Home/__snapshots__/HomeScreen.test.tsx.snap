// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen to match snapshot 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      {
        "paddingHorizontal": 25.142857142857142,
      }
    }
  >
    <Text>
      Last Visited Contact
    </Text>
    <RCTScrollView
      contentContainerStyle={
        {
          "gap": 31.428571428571427,
        }
      }
      data={
        [
          {
            "age": 25,
            "callHistory": "2024-06-09 14:00:00",
            "firstName": "Jane",
            "id": "2",
            "lastName": "Doe",
            "photo": "https://randomuser.me/api/portraits/women/1.jpg",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      showsHorizontalScrollIndicator={false}
      stickyHeaderIndices={[]}
      style={
        {
          "marginVertical": 25.142857142857142,
        }
      }
      testID="list-last-visited-contact"
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
              }
            }
          >
            <View
              style={
                {
                  "borderRadius": 39.285714285714285,
                  "height": 78.57142857142857,
                  "width": 78.57142857142857,
                }
              }
            >
              <Image
                source={
                  {
                    "uri": "https://randomuser.me/api/portraits/women/1.jpg",
                  }
                }
                style={
                  [
                    {
                      "height": "100%",
                      "width": "100%",
                    },
                    {
                      "borderRadius": 39.285714285714285,
                    },
                  ]
                }
              />
            </View>
            <Text>
              Jane Doe
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "backgroundColor": "#D9D9D9",
          "height": 1,
          "width": "100%",
        }
      }
    />
    <RCTScrollView
      data={
        [
          {
            "age": 30,
            "callHistory": "2024-06-09 12:00:00",
            "firstName": "John",
            "id": "1",
            "lastName": "Doe",
            "photo": "https://randomuser.me/api/portraits/men/1.jpg",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomWidth": 1,
                "borderColor": "#D9D9D9",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "opacity": 1,
                "padding": 16,
              }
            }
            testID="contact-card"
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "gap": 23.57142857142857,
                }
              }
            >
              <View
                style={
                  [
                    false,
                    {
                      "overflow": "hidden",
                    },
                    false,
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    [
                      false,
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "backgroundColor": "#D9D9D9",
                        "borderRadius": 40.85714285714286,
                        "height": 81.71428571428572,
                        "width": 81.71428571428572,
                      }
                    }
                  >
                    <Image
                      source={
                        {
                          "uri": "https://randomuser.me/api/portraits/men/1.jpg",
                        }
                      }
                      style={
                        {
                          "borderRadius": 40.85714285714286,
                          "height": "100%",
                          "width": "100%",
                        }
                      }
                      testID="img-contact-card"
                    />
                  </View>
                </View>
              </View>
              <View
                style={
                  {
                    "justifyContent": "space-between",
                  }
                }
              >
                <View>
                  <View
                    style={
                      [
                        false,
                        {
                          "overflow": "hidden",
                        },
                        false,
                        undefined,
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          false,
                          undefined,
                        ]
                      }
                    >
                      <Text
                        style={
                          {
                            "color": "#26245D",
                            "flexShrink": 1,
                            "fontFamily": "Roboto-Bold",
                            "fontSize": 22,
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          }
                        }
                        testID="contact-card-text-name"
                      >
                        John Doe
                      </Text>
                    </View>
                  </View>
                  <View
                    style={
                      [
                        false,
                        {
                          "overflow": "hidden",
                        },
                        false,
                        undefined,
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          false,
                          undefined,
                        ]
                      }
                    >
                      <Text
                        style={
                          {
                            "color": "#5C81F5",
                            "flexShrink": 1,
                            "fontFamily": "Roboto-Bold",
                            "fontSize": 15.714285714285714,
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          }
                        }
                        testID="contact-card-text-age"
                      >
                        30 Years Old
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    [
                      false,
                      {
                        "overflow": "hidden",
                      },
                      false,
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      [
                        false,
                        undefined,
                      ]
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#898898",
                          "flexShrink": 1,
                          "fontFamily": "Roboto-Regular",
                          "fontSize": 15.714285714285714,
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        }
                      }
                      testID="contact-card-call-history"
                    >
                      June 09, 2024
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                [
                  false,
                  {
                    "overflow": "hidden",
                  },
                  false,
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    false,
                    undefined,
                  ]
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#000000",
                        "fontSize": 25.142857142857142,
                      },
                      undefined,
                      {
                        "fontFamily": "anticon",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#5C81F5",
        "borderRadius": 30,
        "bottom": 90,
        "height": 94.28571428571428,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 31.500000000000004,
        "width": 94.28571428571428,
      }
    }
    testID="add-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#ffffff",
            "fontSize": 94.28571428571428,
          },
          undefined,
          {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
  </View>
</RCTSafeAreaView>
`;
